bindings/rust-examples/client-hello-config-resolution/src/bin/async_load_server.rs

```diff
@@ -0,0 +1,132 @@
+        AsyncAnimalConfigResolver { cert_directory }
+    }
+
+    // This method will lookup the appropriate certificates and read them from disc
+    // in an async manner which won't block the tokio task.
+    async fn server_config(
+        &self,
+        animal: String,
+    ) -> Result<s2n_tls::config::Config, s2n_tls::error::Error> {
+        let cert_path = format!("{}/{}-chain.pem", self.cert_directory, animal);
+        let key_path = format!("{}/{}-key.pem", self.cert_directory, animal);
+        // we asynchronously read the cert chain and key from disk
+        let (cert, key) = try_join!(tokio::fs::read(cert_path), tokio::fs::read(key_path))
+            // we map any IO errors to the s2n-tls Error type, as required by the ConfigResolver bounds.
+            .map_err(|io_error| s2n_tls::error::Error::application(Box::new(io_error)))?;
+
+        let mut config = s2n_tls::config::Builder::new();
+        // we can set different policies for different configs. "20190214" doesn't
+        // support TLS 1.3, so any customer requesting www.wombat.com won't be able
+        // to negotiate TLS 1.3
+        let security_policy = match animal.as_str() {
+            "wombat" => Policy::from_version("20190214")?,
+            _ => DEFAULT_TLS13,
+        };
+        config.set_security_policy(&security_policy)?;
+        config.load_pem(&cert, &key)?;
+        config.build()
+    }
+}
+
+impl ClientHelloCallback for AsyncAnimalConfigResolver {
+    fn on_client_hello(
+        &self,
+        connection: &mut s2n_tls::connection::Connection,
+    ) -> Result<Option<Pin<Box<dyn ConnectionFuture>>>, s2n_tls::error::Error> {
+        let sni = match connection.server_name() {
+            Some(sni) => sni,
+            None => {
+                println!("connection contained no SNI");
+                return Err(s2n_tls::error::Error::application("no sni".into()));
+            }
+        };
+
+        // simple, limited logic to parse "animal" from "www.animal.com".
+        let mut tokens = sni.split('.');
+        tokens.next(); // "www"
+        let animal = match tokens.next() {
+            Some(animal) => animal.to_owned(), // "animal"
+            None => {
+                println!("unable to parse sni");
+                return Err(s2n_tls::error::Error::application(
+                    format!("unable to parse sni: {}", sni).into(),
+                ));
+            }
+        };
+
+        // A ConfigResolver can be constructed from a future that returns
+        // `Result<Config, s2n_tls::error::Error>`, with the main additional
+        // requirements that the future is `'static`, which generally means that
+        // it can't have any  interior references. This will prevent you from
+        // doing something like
+        // ```
+        // let config_resolver = ConfigResolver::new(self.server_config(animal));
+        // ```
+        // because the compiler will complain that `&self` doesn't live long enough.
```

Comment:
Yes on the first point.

Implementing `Future` it a pretty ugly way to go, because then you have to write the poll method yourself, manage wakers, etc, all of which is very unpleasant.

I did try and simplify it a bit by just having to `server_config` _consume_ self. That way the future explicitly owns `self` and no `async move` shenanigans are required.

Reply:
Oh interesting! That's cool!